src/miniscript/iter.rs

```diff
@@ -0,0 +1,552 @@
+            _ => vec![],
+        }
+    }
+
+    /// Returns `Vec` with hashes of all public keys from the current miniscript item, if any.
+    /// Otherwise returns an empty `Vec`.
+    ///
+    /// For each public key the function computes hash; for each hash of the public key the function
+    /// returns it's cloned copy.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    /// To obtain a list of all public key hashes within AST use [`iter_pkh()`] function,
+    /// for example `miniscript.iter_pubkey_hashes().collect()`.
+    pub fn get_leaf_pkh(&self) -> Vec<Pk::Hash> {
+        match self.node.clone() {
+            Terminal::PkH(hash) => vec![hash],
+            Terminal::PkK(key) => vec![key.to_pubkeyhash()],
+            Terminal::Multi(_, keys) => keys.iter().map(Pk::to_pubkeyhash).collect(),
+            _ => vec![],
+        }
+    }
+
+    /// Returns `Vec` of [PubkeyOrHash] entries, representing either public keys or public key
+    /// hashes, depending on the data from the current miniscript item. If there is no public
+    /// keys or hashes, the function returns an empty `Vec`.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    /// To obtain a list of all public keys or hashes within AST use [`iter_pk_pkh()`]
+    /// function, for example `miniscript.iter_pubkeys_and_hashes().collect()`.
+    pub fn get_leaf_pk_pkh(&self) -> Vec<PkPkh<Pk>> {
+        use self::PkPkh::*;
+        match self.node.clone() {
+            Terminal::PkH(hash) => vec![HashedPubkey(hash)],
+            Terminal::PkK(key) => vec![PlainPubkey(key)],
+            Terminal::Multi(_, keys) => keys.into_iter().map(PlainPubkey).collect(),
+            _ => vec![],
+        }
+    }
+}
+
+/// Iterator for traversing all [Miniscript] miniscript AST references starting from some specific
+/// node which constructs the iterator via [Miniscript::iter] method.
+pub struct Iter<'a, Pk: 'a + MiniscriptKey, Ctx: 'a + ScriptContext> {
+    next: Option<&'a Miniscript<Pk, Ctx>>,
+    path: Vec<(&'a Miniscript<Pk, Ctx>, usize)>,
+}
+
+impl<'a, Pk: MiniscriptKey, Ctx: ScriptContext> Iter<'a, Pk, Ctx> {
+    fn new(miniscript: &'a Miniscript<Pk, Ctx>) -> Self {
+        Iter {
+            next: Some(miniscript),
+            path: vec![],
+        }
+    }
+}
+
+impl<'a, Pk: MiniscriptKey, Ctx: ScriptContext> Iterator for Iter<'a, Pk, Ctx> {
+    type Item = &'a Miniscript<Pk, Ctx>;
+
+    /// First, the function returns `self`, then the first child of the self (if any),
+    /// then proceeds to the child of the child — down to a leaf of the tree in its first branch.
+    /// When the leaf is reached, it goes in the reverse direction on the same branch until it
+    /// founds a first branching node that had more than a single branch and returns it, traversing
+    /// it with the same algorithm again.
```

Comment:
I'm having trouble understanding this comment. Do you just mean that it does an in-order traversal of the entire tree?

Reply:
Yes, this is what was meant